### PR TITLE
Fix issue #446: Add 'duration' selector on project request 

### PR DESCRIPTION
**Pull request by rsiminel**
Is this sufficient for fixing the issue ?
As I understand, an admin is supposed to create each project manually anyway, copying the information off of the email they were sent. This means that now, they'll have to copy the expiration date manually as well. Is there a better way of doing things when add_project (from manager2/src/app/admin/projects.component.ts) is executed ?

### CODE REVIEW

**Review by mboudet — COMMENTED** [review]
Seems fine on the 'request' side. However, we want the selected end date to be available on the admin side for validation.

On the [admin side](https://github.com/genouest/genouestaccountmanager/blob/364e137b3138e132539f3965618286ed0d179332/manager2/src/app/admin/project/project.component.html#L47), the column/parameter is called 'expire'. it need to match the column/parameter you are storing in DB (currently 'end')

File: manager2/src/app/project/project.component.html [hunk 1]
```diff
@@ -29,6 +29,12 @@ <h3>Request a new project creation</h3>
               <small id="nameHelp" class="form-text text-muted"><i>Optional, for information only</i></small>
           </div>
 
+          <div class="form-group row col-sm-8">
+            <label for="new_project_end">Expected end date</label>
+            <input placeholder="project end date" type="date" id="new_project_end" [ngModelOptions]="{standalone: true}" [(ngModel)]="new_project.end" class="form-control"/>
+            <small id="nameHelp" class="form-text text-muted"><i>Optional, for information only</i></small>
```
[the remark this answers is not in the document]
Comment (rsiminel) [reply manager2/src/app/project/project.component.html]: Ah, I had not understood that. On it.

[the remark this answers is not in the document]
Comment (rsiminel) [reply manager2/src/app/project/project.component.html]: Did that do the trick ?
I'm not sure how the security over DataTime formating works in TypeScript.
I asume we should also modify line 26 in project.service.js in /core ?
How do we allow the admin validating the request to manually modify the expiration date in case the user entered something stupid ?

[the remark this answers is not in the document]
Comment (mboudet) [reply manager2/src/app/project/project.component.html]: This should work, yeah.
Before 'validating' a project, an administrator can modify all fields, so we can correct any issues with provided values.

File: manager2/src/app/admin/projects/projects.component.ts [hunk 2]
```diff
@@ -175,15 +175,15 @@ export class ProjectsComponent implements OnInit {
         this.projectService.add({
             'uuid': this.new_project.uuid,
             'id': this.new_project.id,
+            'expire': this.new_project.expire,
```
Comment (mboudet) [inline manager2/src/app/admin/projects/projects.component.ts:178]: I 'think' you are going to need to convert date several times (in this file, but also on the user project file).

The issue is that angular datepicker give you a Date (a js entity type). But we don't want to store it in DB as it is, we want to store the unix time

(Using the `new Date(this.new_project.expire).getTime()` there was previously).
But then, you will need to convert it back to display it on the admin side (using something like [this](https://github.com/genouest/genouestaccountmanager/blob/364e137b3138e132539f3965618286ed0d179332/manager2/src/app/admin/project/project.component.ts#L190))

To summarize, the process would be:

* User select date in the field, and click send
* Before sending to the service, you convert the value using `new Date(this.new_project.expire).getTime()` (probably should not edit the variable in-place. Might be better to pass a whole new project Object with the values you want)
* Store the project in DB as you already do.
* When an admin access the projects admin page, get the project from DB, and convert it back to date from display

Comment (rsiminel) [reply manager2/src/app/admin/projects/projects.component.ts]: It should be good now.

Comment (mboudet) [reply manager2/src/app/admin/projects/projects.component.ts]: Still need to convert it to unix time on the 'request project' side, and convert it back to JS Date when displaying admin-side

Comment (rsiminel) [reply manager2/src/app/admin/projects/projects.component.ts]: Right, sorry for taking so long, I had to figure out what the heck a spread operator is.
Is there anything else I missed?

Comment (mboudet) [reply manager2/src/app/admin/projects/projects.component.ts]: Hmm, I think you still need to convert unix time to JS date on the admin side when getting the pending projets
Now that I'm thinking about it, you'll also need to manage the case when pending projects do not have an expire data (since current pending projects in production will not have one)

Also need to rollback package-lock.js and dev_quickstart/default.json I think?

Comment (rsiminel) [reply manager2/src/app/admin/projects/projects.component.ts]: Ok so Angular's [(ngModel)] directive handles conversion between Unix timestamps and date objects automatically so there is no need to do admin-side conversions to JS date types but we do have to convert back to unix timestamps manually.
Also, the GitLens extention for VS Code is fantastic. Here you are working on the same problem 5 years ago : https://github.com/genouest/genouestaccountmanager/pull/33

Comment (rsiminel) [reply manager2/src/app/admin/projects/projects.component.ts]: Ok, @mboudet, everything works now. The stupid html date input controls only accept a formated string (which is the original reason why the date_convert() function was written), which was causing me issues.

If there is a pending project with no expiration date, it will not cause any bugs but the admin accepting the project will have to enter one manually.

File: manager2/src/app/admin/projects/projects.component.ts [hunk 3]
```diff
@@ -147,10 +147,6 @@ export class ProjectsComponent implements OnInit {
         // about order, see: https://medium.com/@lukaonik/how-to-fix-the-previous-ngmodelchange-previous-value-in-angular-6c2838c3407d
         this.new_project.id = tmpprojectid; // todo: maybe add an option to enable or disable this one
 
-        if (!this.new_project.expire) {
-            this.new_project.expire = this.date_convert(new Date().getTime() + this.config.project.default_expire * this.day_time)
-        }
-
```
Comment (mboudet) [inline manager2/src/app/admin/projects/projects.component.ts:153]: Why not keep this as a fallback if expire is not set (old pending projects), instead of being an empty string?

Comment (rsiminel) [reply manager2/src/app/admin/projects/projects.component.ts]: You're right, I brought it back.
While I'm at it, may I ask if [this](https://github.com/rsiminel/genouestaccountmanager/blob/56aca6f198bf2c5bce5a26d2099373424f103ff5/manager2/src/app/admin/project/project.component.ts#L92) is a sanctioned use of date_convert ?

Comment (mboudet) [reply manager2/src/app/admin/projects/projects.component.ts]: Not sure what you mean by 'sanctioned use' ?

Comment (rsiminel) [reply manager2/src/app/admin/projects/projects.component.ts]: Sorry, I was unclear. I meant to ask why the function is being called there and whether it might cause the date to be stored as a string in the database after that function call.

Comment (mboudet) [reply manager2/src/app/admin/projects/projects.component.ts]: I think 'show_project_users' is in charge of actually getting the project. Since we show the expiration date in a editable field, I believe it need to be converted from DB format in that step.

We convert it back before updating [here](https://github.com/genouest/genouestaccountmanager/blob/364e137b3138e132539f3965618286ed0d179332/manager2/src/app/admin/project/project.component.ts#L164)

Comment (rsiminel) [reply manager2/src/app/admin/projects/projects.component.ts]: Oh ok, thanks for explanation. I guess we can merge now?

**Comment by mboudet** [comment]
Looks good to me. Does everything behave as expected?

**Review by mboudet — APPROVED** [review]
